### PR TITLE
[FLINK-1415] Akka cleanups

### PR DESCRIPTION
**Pull request by tillrohrmann**
This PR removes the akka.jobmanager.url config parameter to override the jobmanager address and port set in the configuration.

Adds timeout heuristics to deduce the different timeout values from a single value if no other values are specified. This makes it easier to configure the system.

Adds test cases to test if the JobManager detects a failing TaskManager using Akka's death watch. Moreover, the test cases test alos if the TaskManager detects a failing JobManager and then tries to reconnect to it.

Removes the notifyExecutionStateChange method which gave access to the internal actor state of the TaskManager. Replaced by sending a message to the TaskManager. This removes the NullPointerExceptions which occurred when shutting the system down.

Harmonizes the scala code and adds comments to the important classes.

This PR is based on the PR #317, which is required by the refactoring of the notifyExecutionStateChange method.

### CODE REVIEW

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/SlotAvailabilityListener.java [hunk 1]
```diff
@@ -21,7 +21,7 @@
 import org.apache.flink.runtime.instance.Instance;
 
 /**
- * A SlotAvailabilityListener can be notified when new {@link org.apache.flink.runtime.instance.AllocatedSlot}s become available
+ * A SlotAvailabilityListener can be notified when new {@link org.apache.flink.runtime.instance.AllocatedSlot2}s become available
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/SlotAvailabilityListener.java]: I guess `AllocatedSlot2` is an automatic rename leftover

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/SlotSharingGroupAssignment.java [hunk 2]
```diff
@@ -42,86 +44,82 @@
...
-	}
-	
-	private SubSlot addNewSlotWithTask(AllocatedSlot slot, AbstractID groupId, JobVertexID vertexId) {
-		
-		final SharedSlot sharedSlot = new SharedSlot(slot, this);
-		final Instance location = slot.getInstance();
-		
+
+	public SimpleSlot addSharedSlotAndAllocateSubSlot(SharedSlot sharedSlot, Locality locality,
+													AbstractID groupId, CoLocationConstraint constraint) {
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/SlotSharingGroupAssignment.java]: indentation?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 3]
```diff
@@ -329,6 +330,15 @@ public void unregisterMemoryManager(MemoryManager memoryManager) {
 		}
 	}
 
+	protected void notifyExecutionStateChange(ExecutionState executionState,
+											Throwable optionalError) {
```
Comment (rmetzger) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]: This also seems weird

**Comment by rmetzger** [comment]
This is a huge change. I took a look over the code, but I don't have enough experience with the scheduler to understand these changes.

I would suggest to merge this rather soon because its touching a lot of code due to minor scala style changes (semicolons, removal of parentheses from no-arg methods, unneeded { } and so on)
+1 for the added documentation to the classes!

The bug in FLINK-1453 would be more obvious with these changes were merged. That's another motivation for me to push this pull request forward ;)

**Comment by uce** [comment]
I think @StephanEwen is reviewing the critical part.

**Comment by tillrohrmann** [comment]
I think so too. The changes of the scheduler are only included because otherwise the test cases wouldn't pass. The scheduler relevant PR is #317.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/SlotAvailabilityListener.java (hunk 1 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/SlotAvailabilityListener.java]: You are right. Good catch. I'll correct it.

File: flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/SlotSharingGroupAssignment.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/jobmanager/scheduler/SlotSharingGroupAssignment.java]: I must have fallen asleep on the space button.

**Comment by rmetzger** [comment]
Do you think this change is also addressing this error?
https://travis-ci.org/rmetzger/flink/jobs/48365538

**Comment by tillrohrmann** [comment]
To be honest, I have not seen this error before. But I doubt that this is a
subslot release issue. If this error appears more often, then we could try
to increase the start up timeout in case that it is a travis resource issue.

**Comment by tillrohrmann** [comment]
Rebased on Stephan's latest master branch containing the subslot release fix and reduced memory footprint of archived ExecutionGraphs.

**Pull request closed without merging**